src/dict.js

```diff
@@ -157,47 +182,94 @@ Sk.builtin.dict.prototype.sq$contains = function (ob) {
 };
 
 Sk.builtin.dict.prototype.mp$ass_subscript = function (key, w) {
-    var k = kf(key);
-    var bucket = this.buckets[k.v];
-    var item;
+    const hash = kf(key);
+    let item;
+    if (typeof hash === "string") {
+        // we have a string so add to entries directly
+        item = this.entries[hash];
+        if (item === undefined) {
+            this.entries[hash] = { lhs: key, rhs: w };
+            this.size += 1;
+        } else {
+            item.rhs = w;
+        }
+    } else {
+        item = this.get$bucket_item(key, hash);
+        if (item === undefined) {
+            this.set$bucket_item(key, w, hash);
+            this.size += 1;
+        } else {
+            item.rhs = w;
+        }
+    }
+};
 
+Sk.builtin.dict.prototype.set$bucket_item = function (key, value, hash_value) {
+    // put a key, value pair into appropriate bucket and insert into entries
+    let key_hash,
+        bucket = this.buckets[hash_value];
+    const item = { lhs: key, rhs: value };
     if (bucket === undefined) {
-        // New bucket
-        bucket = {$hash: k, items: [
-            {lhs: key, rhs: w}
-        ]};
-        this.buckets[k.v] = bucket;
-        this.size += 1;
-        return;
+        bucket = this.buckets[hash_value] = [];
+        key_hash = "#" + hash_value + "_" + 0; // this is the zeroth entry
+        bucket.push(item);
+    } else {
+        // we might have a freeslot from deleting an item
+        // so either insert into freeslot or push
+        const free_slot_idx = bucket.indexOf(undefined);
+        if (free_slot_idx !== -1) {
+            key_hash = "#" + hash_value + "_" + free_slot_idx;
+            bucket[free_slot_idx] = item;
+        } else {
+            key_hash = "#" + hash_value + "_" + bucket.length;
+            bucket.push(item);
+        }
     }
+    this.entries[key_hash] = item;
+};
 
-    item = this.key$lookup(bucket, key);
-    if (item) {
-        item.rhs = w;
-        return;
-    }
 
-    // Not found in dictionary
-    bucket.items.push({lhs: key, rhs: w});
-    this.size += 1;
+
+
+Sk.builtin.dict.prototype.pop$bucket_item = function (key, hash_value) {
+    // pop a key, value pair and remove reference in buckets and entries
+    const bucket = this.buckets[hash_value];
+    let bucket_key, item;
+    if (bucket === undefined) {
+        return undefined;
+    }
+    for (let i = 0; i < bucket.length; i++) {
+        item = bucket[i];
+        if (item === undefined) {
+            continue;
+        }
+        bucket_key = item.lhs;
+        if (bucket_key === key || Sk.misceval.richCompareBool(key, bucket_key, "Eq")) {
+            const key_hash = "#" + hash_value + "_" + i;
+            delete this.entries[key_hash];
+            bucket[i] = undefined; // undefined signals this slot is free for use in the event of reinsertion/collision
+            return item;
+        }
+    }
+    return;
```

Comment:
consider cleaning up `this.buckets[hash_value]` if `bucket.length == 0`

Reply:
@LeszekSwirski  i'm not sure i follow...

Reply:
If `bucket.length == 0`, then afaict you can `delete this.buckets[hash_value]`. If I'm mistaken then please ignore me :)

Reply:
do you mean something like? 

```javascript
    const bucket = this.buckets[hash_value];
    let bucket_key, item;
    if (bucket === undefined || bucket.length === 0) {
        return;
    }
```
I still think I'm missing what you mean... sorry

Reply:
Oh I think i see now - are you saying if this was the only `item` in the `bucket` then we can `delete` the `bucket` it... rather than setting it to `undefined` (which we'd have to do if there were multiple items)??

so something like

```javascript
if (bucket.length === 1) {
    delete this.buckets[hash_value];
} else {
    bucket[i] = undefined; // undefined signals this slot is free for use in the event of reinsertion/collision
}
return item

```

Reply:
What I mean is, that if all the entries of bucket are undefined (I guess that's more complex than just bucket.length being set to zero, you may have to also do `delete bucket[i]` rather than setting undefined), then you can delete the entire bucket from the set of buckets, i.e. do the inverse of line 213. Basically just a memory cleanup for empty buckets. It's not a big deal probably :)

Reply:
ok I think i've got it... maybe